pallets/circuit/src/square_up/mod.rs

```diff
@@ -0,0 +1,214 @@
+
+        Ok(())
+    }
+
+    /// Fallible bidding attempt by executors.
+    /// Input: LocalXtxCtx, bidder, bid_amount, bid_asset_id
+    /// Output: Result<(), Error<T>>
+    pub fn try_bid(
+        sfx_id: T::Hash,
+        requester: &T::AccountId,
+        bidder: &T::AccountId,
+        bid: &SFXBid<T::AccountId, BalanceOf<T>, u32>,
+        current_best_bid: Option<SFXBid<T::AccountId, BalanceOf<T>, u32>>,
+    ) -> DispatchResult {
+        let total_bid_deposit = bid
+            .amount
+            .checked_add(&bid.reserved_bond.unwrap_or(Zero::zero()))
+            .ok_or(Error::<T>::ArithmeticErrorOverflow)?
+            .checked_add(&bid.insurance)
+            .ok_or(Error::<T>::ArithmeticErrorOverflow)?;
+
+        match current_best_bid {
+            Some(current_best_bid) => {
+                if bid.amount >= current_best_bid.amount {
+                    return Err(Error::<T>::BiddingRejectedBetterBidFound.into())
+                }
+                <T as Config>::AccountManager::transfer_deposit(
+                    current_best_bid.generate_id::<SystemHashing<T>, T>(sfx_id),
+                    bid.generate_id::<SystemHashing<T>, T>(sfx_id),
+                    Some(total_bid_deposit),
+                    Some(&bid.executor),
+                    None,
+                )
+            },
+            None => <T as Config>::AccountManager::deposit(
+                bid.generate_id::<SystemHashing<T>, T>(sfx_id),
+                RequestCharge {
+                    payee: bidder.clone(),
+                    offered_reward: total_bid_deposit,
+                    charge_fee: Zero::zero(),
+                    source: BenefitSource::TrafficRewards,
+                    role: CircuitRole::Executor,
+                    recipient: Some(requester.clone()),
+                    maybe_asset_id: bid.reward_asset_id,
+                },
+            ),
+        }
+    }
+
+    /// Infallible re-balance requesters locked rewards after possibly lower bids are posted.
+    pub fn bind_bidders(local_ctx: &LocalXtxCtx<T>) {
+        for fsx in Machine::<T>::read_current_step_fsx(local_ctx).iter() {
+            let sfx_id = fsx.generate_id::<SystemHashing<T>, T>(local_ctx.xtx_id);
+            if let Some(bid) = &fsx.best_bid {
+                if !<T as Config>::AccountManager::assign_deposit(sfx_id, &bid.executor) {
+                    log::error!(
+                        "assign_deposit: expect assign_deposit to succeed for sfx_id: {:?}",
+                        sfx_id
+                    );
+                }
+            } else {
+                log::error!(
+                    "bind_bidders: expect best_bid to be Some for sfx_id: {:?}",
+                    sfx_id
+                );
+            }
+        }
+    }
+
+    /// Drop Xtx and unlock requester and all executors that posted bids - without penalties.
+    pub fn kill(local_ctx: &LocalXtxCtx<T>) {
```

Comment:
kill wouldn't cancel_deposits only if called with incorrect SFX ID to cancel the requester's deposit or BID ID  is incorrect to cancel bidder's deposit. There will be a pending RequestCharge entry in account manager as well as "kill: expect cancel_deposit to succeed for sfx_id: {:?}",err log messages. 

I added a manual cancel_xtx operation to be called with ID in case the ids get messed up causing RequestCharges to remain so we could kick it manually. 

I could open a ticket to populate an "emergency" storage entries for failed cancel_deposits attempt so that we could log them for inspection?

Reply:
Good idea, as long as we have ways to recover this so it's not locked, I think it's fine. We should think about this "retryability" going forward either as A. Passive retryability (the system recovers itself) or B. Active retryability(a user can recover themselves).